### PR TITLE
convert report timeseries to plotly

### LINKED ISSUE #365: Update report timeseries figures to plotly
So that pdf creation for report pdf generation is simpler and to reduce the number if libraries a user has to load. 
Should investigate to see if this should be included in beta5 or not.

### PR DESCRIPTION
**Pull request by lboeman**
- [x] Closes #365 .
  - [x] I am familiar with the [contributing guidelines](https://solarforecastarbiter-core.readthedocs.io/en/latest/contributing.html).
  - [x] Tests added.
  - [x] Updates entries to [`docs/source/api.rst`](https://github.com/SolarArbiter/solarforecastarbiter-core/blob/master/docs/source/api.rst) for API changes.
  - [x] Adds descriptions to appropriate "what's new" file in [`docs/source/whatsnew`](https://github.com/SolarArbiter/solarforecastarbiter-core/tree/master/docs/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
  - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
  - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

Converts the timeseries and scatterplots of reports to plotly.

### CODE REVIEW

**Comment by wholmgren** [comment]
closes #253? is it possible to close #351 in this PR too? 

Probably will need a follow up issue and PR for the observation figures with the data quality flag bars.

**Comment by lboeman** [comment]
> closes #253? is it possible to close #351 in this PR too?
> 
> Probably will need a follow up issue and PR for the observation figures with the data quality flag bars.

The legend has the same hiding behavior as the bokeh plots. So I'm not sure if this closes 253. 

I can definitely update this to also close 351, I poked around at the functionaility while working on it and it seems like a simple update, with a couple of potential gotchas that I can take a look at early.

**Comment by wholmgren** [comment]
> The legend has the same hiding behavior as the bokeh plots. So I'm not sure if this closes 253.

double clicking an item in the legend hides all other traces. I noticed this when using the metrics plots and just confirmed that it works for the time series plot you shared over slack last week.

**Comment by lboeman** [comment]
> double clicking an item in the legend hides all other traces. I noticed this when using the metrics plots and just confirmed that it works for the time series plot you shared over slack last week. So this would close #253 as I envisioned it.


Well, that's really cool. I didn't realize that. huzzah.

**Comment by lboeman** [comment]
Added a `_fill_timeseries` function to create gaps in plotted timeseries data addressing #351. The gaps appear as below:
Screenshot from 2020-03-30 12-06-27

**Comment by wholmgren** [comment]
nice!

**Comment by lboeman** [comment]
@wholmgren Should this have a more descriptive entry in the docs? It's got a basic "changed plots to plotly" entry, but perhaps noting any new features/ issues closed is a good idea here?

**Comment by wholmgren** [comment]
Yeah, should note the issues that were closed along the way. It'll help me next week when writing the dreaded quarterly report.

File: solarforecastarbiter/reports/figures/plotly_figures.py [hunk 1]
```diff
@@ -35,9 +39,320 @@
...
+                 pfxobs.interval_label))),
+            'observation_color': _obs_color(
+                pfxobs.interval_length)
+        }
+        value_frames.append(pd.DataFrame(value_frame_dict))
+        meta_rows.append(meta_row_dict)
+    data = pd.concat(value_frames)
+    metadata = pd.DataFrame(meta_rows)
+    # drop tz info from localized times. GH164
+    data = data.tz_localize(None)
```
Comment (alorenzo175) [inline solarforecastarbiter/reports/figures/plotly_figures.py]: Is this still necessary? I think it was required for bokeh

File: solarforecastarbiter/reports/figures/plotly_figures.py [hunk 2]
```diff
@@ -35,9 +39,320 @@
...
+def _fill_timeseries(df, interval_length):
+    """Returns a dataframe with a datetimeindex with regular frequency of
+    interval_length minutes. Previously missing values will be filled with
+    nans. Useful for creating gaps in plotted timeseries data.
+
+    Parameters
+    ----------
+    df: pandas.DataFrame
+        Dataframe with timeseries data.
+    interval_length: timedelta-li
```
Comment (alorenzo175) [inline solarforecastarbiter/reports/figures/plotly_figures.py]:
```suggestion
    interval_length: timedelta-like
```

File: solarforecastarbiter/reports/figures/plotly_figures.py [hunk 3]
```diff
@@ -35,9 +39,320 @@
...
+        Interval length of the data in minutes.
```
Comment (alorenzo175) [inline solarforecastarbiter/reports/figures/plotly_figures.py]: not necessarily in minutes no?

File: solarforecastarbiter/reports/figures/plotly_figures.py [hunk 4]
```diff
@@ -35,9 +39,320 @@
...
+    obs_color = OBS_PALETTE[idx]
+    return obs_color
+
+
+def _boolean_filter_indices_by_pair(value_cds, pair_index):
+    return value_cds.data['pair_index'] == pair_index
+
+
+def _extract_metadata_from_df(metadata_df, hash_, hash_key):
+    # first_row = np.argwhere(metadata_df[hash_key] == hash_)[0][0]
```
Comment (alorenzo175) [inline solarforecastarbiter/reports/figures/plotly_figures.py]: delete

File: solarforecastarbiter/reports/template.py [hunk 5]
```diff
@@ -75,17 +75,13 @@ def _get_render_kwargs(report, dash_url, with_timeseries):
     kwargs['plotly_version'] = plot_plotly if plot_plotly else plotly_version
     if with_timeseries:
         try:
-            script, div = bokeh_figures.timeseries_plots(report)
+            timeseries_specs = plotly_figures.timeseries_plots(report)
         except Exception:
             logger.exception(
                 'Failed to make Bokeh items for timeseries and scatterplot')
```
Comment (alorenzo175) [inline solarforecastarbiter/reports/template.py]:
```suggestion
                'Failed to make Plotly items for timeseries and scatterplot')
```

File: solarforecastarbiter/reports/figures/plotly_figures.py [hunk 6]
```diff
@@ -35,9 +39,320 @@
...
+    end : pandas.Timestamp
+        Report end time
+    timezone : str
+        Timezone consistent with the data in the timeseries_metadata_df.
+
+    Returns
+    -------
+    plotly.Figure
+    """  # NOQA
+    palette = cycle(PALETTE)
```
Comment (alorenzo175) [inline solarforecastarbiter/reports/figures/plotly_figures.py]: lgtm says this is unnecessary

File: solarforecastarbiter/reports/figures/plotly_figures.py [hunk 7]
```diff
@@ -35,9 +39,320 @@
...
+    timeseries_meta_df: pandas.DataFrame
+        DataFrame of metadata for each Observation Forecast pair. See
+        :py:func:`solarforecastarbiter.reports.figures.construct_timeseries_dataframe`
+        for format.
+
+    Returns
+    -------
+    plotly.Figure
+    """  # NOQA
+    xy_min, xy_max = _get_scatter_limits(timeseries_value_df)
```
Comment (alorenzo175) [inline solarforecastarbiter/reports/figures/plotly_figures.py]: are these not used? should they be?

**Comment by lboeman** [comment]
> can we also remove bokeh from head.html and template.py?

I maintained the import of the color palette from bokeh. How do you feel about just manually constructing this list and hard coding it as a global in the plotly_figures.py module? 

Not sure if there's an agreeable palette supplied by plotly. [Plotly color swatches](https://plotly.com/python/discrete-color/#color-sequences-in-plotly-express)

**Comment by wholmgren** [comment]
I'm fine with hard coding. I like most of plotly's 10 color palettes, but beyond that they are pretty bad. The d3 category 20 palette is about as good as it gets with that many colors. At any rate I think it's better than only 10 colors or ugly colors. Better still would be to apply some other kind of styling when the number of colors is greater than 10 but that was more work than I wanted to put into it.

**Comment by alorenzo175** [comment]
a hardcoded palette is fine by me

**Comment by lboeman** [comment]
Alright, I've hardcoded the d3 category 20 palette in there and created a function to generate grayscale similar to the bokeh one we were using.

**Comment by lboeman** [comment]
> bokeh script loading in head.html?

Thanks!

**Pull request merged**